api/saml/provider.py

```diff
@@ -0,0 +1,335 @@
+        }
+
+        configuration_serializer = SAMLConfigurationSerializer(self.external_integration(db))
+        metadata_serializer = SAMLMetadataSerializer(self.external_integration(db))
+        configuration = SAMLConfiguration(configuration_serializer, metadata_serializer)
+
+        for identity_provider in configuration.identity_providers:
+            link = {
+                'rel': 'authenticate',
+                'href': self._create_authenticate_url(db, identity_provider.entity_id),
+                'display_names': self._join_ui_info_items(identity_provider.ui_info.display_names),
+                'descriptions': self._join_ui_info_items(identity_provider.ui_info.descriptions),
+                'information_urls': self._join_ui_info_items(identity_provider.ui_info.information_urls),
+                'privacy_statement_urls': self._join_ui_info_items(
+                    identity_provider.ui_info.privacy_statement_urls),
+                'logo_urls': self._join_ui_info_items(identity_provider.ui_info.logo_urls)
+            }
+
+            flow_doc['links'].append(link)
+
+        return flow_doc
+
+    def _create_token(self, db, patron, token):
+        """Creates a Credential object that ties the given patron to the
+        given provider token.
+
+        :param db: Database session
+        :type db: sqlalchemy.orm.session.Session
+
+        :param patron: Patron object
+        :type patron: Patron
+
+        :param token: Token containing SAML subject's metadata
+        :type token: Dict
+
+        :return: Credential object
+        :rtype: Credential
+        """
+        data_source, ignore = self._get_token_data_source(db)
+        duration = datetime.timedelta(days=self.token_expiration_days)
+        return Credential.temporary_token_create(
+            db, data_source, self.TOKEN_TYPE, patron, duration, token
+        )
+
+    def _create_authenticate_url(self, db, idp_entity_id):
+        """Returns an authentication link used by clients to authenticate patrons
+
+        :param db: Database session
+        :type db: sqlalchemy.orm.session.Session
+
+        :param idp_entity_id: Entity ID of the IdP which will be used for authentication
+        :type idp_entity_id: string
+
+        :return: URL for authentication using the chosen IdP
+        :rtype: string
+        """
+
+        library = self.library(db)
+
+        return url_for(
+            'saml_authenticate',
+            _external=True,
+            library_short_name=library.short_name,
+            provider=self.NAME,
+            idp_entity_id=idp_entity_id
+        )
+
+    def _get_token_data_source(self, db):
+        """Returns a token data source
+
+        :param db: Database session
+        :type db: sqlalchemy.orm.session.Session
+
+        :return: Token data source
+        :rtype: DataSource
+        """
+        return get_one_or_create(
+            db, DataSource, name=self.TOKEN_DATA_SOURCE_NAME
+        )
```

Comment:
I don't think this is going to work in a situation where a library has multiple SAML authentication mechanisms for its patrons. You'll look up a Credential from this data source but you won't be able to tell which IdP it came from.

This would work if the data source identified the IdP -- logically, the IdP is the "source" of this "data".

Reply:
Of course, it's also possible to say that a library can't have multiple SAML authentication mechanisms -- that's what we say now for Basic Auth authentication mechanisms. But this branch isn't there and I don't think that's where you're going.

Reply:
I was looking at `OAuthAuthenticationProvider` as a reference and it has the same logic in `authenticator.py`:
```python
    def token_data_source(self, _db):
        return get_one_or_create(
            _db, DataSource, name=self.TOKEN_DATA_SOURCE_NAME
        )
```

What is the best way to fix it?

Reply:
Ugh, that's no good. I guess let's leave it as is for now, add a comment, and we'll figure it out when we need to.

Reply:
Added a comment